lib/util/git/private-key.ts

```diff
@@ -5,59 +5,132 @@ import upath from 'upath';
+const sshKeyRegex = regEx(
+  /-----BEGIN ([A-Z ]+ )?PRIVATE KEY-----(.|\n)*?-----END ([A-Z]+ )?PRIVATE KEY-----/,
+);
+
+let gitPrivateKey: PrivateKey | undefined;
+
+abstract class PrivateKey {
+  protected readonly key: string;
+  protected keyId: string | undefined;
+  protected abstract readonly gpgFormat: string;
+
+  constructor(key: string) {
+    this.key = key.trim();
+    addSecretForSanitizing(this.key, 'global');
+    logger.debug(
+      'gitPrivateKey: successfully set (but not yet written/configured)',
+    );
+  }
+
+  async writeKey(): Promise<void> {
+    try {
+      if (!this.keyId) {
+        this.keyId = await this.importKey();
+      }
+      logger.debug('gitPrivateKey: imported');
+    } catch (err) {
+      logger.warn({ err }, 'gitPrivateKey: error importing');
+      throw new Error(PLATFORM_GPG_FAILED);
+    }
   }
-  addSecretForSanitizing(key.trim(), 'global');
-  logger.debug(
-    'gitPrivateKey: successfully set (but not yet written/configured)',
-  );
-  gitPrivateKey = key.trim();
+
+  async configSigningKey(cwd: string): Promise<void> {
+    logger.debug('gitPrivateKey: configuring commit signing');
+    // TODO: types (#22198)
+    await exec(`git config user.signingkey ${this.keyId!}`, { cwd });
+    await exec(`git config commit.gpgsign true`, { cwd });
+    await exec(`git config gpg.format ${this.gpgFormat}`, { cwd });
+  }
+
+  protected abstract importKey(): Promise<string | undefined>;
 }
 
-async function importKey(): Promise<void> {
-  if (keyId) {
-    return;
+class GPGKey extends PrivateKey {
+  protected readonly gpgFormat = 'opengpg';
+
+  protected async importKey(): Promise<string | undefined> {
+    const keyFileName = upath.join(os.tmpdir() + '/git-private-gpg.key');
+    await fs.outputFile(keyFileName, this.key);
+    const { stdout, stderr } = await exec(`gpg --import ${keyFileName}`);
+    logger.debug({ stdout, stderr }, 'Private key import result');
+    await fs.remove(keyFileName);
+    return `${stdout}${stderr}`
+      .split(newlineRegex)
+      .find((line) => line.includes('secret key imported'))
+      ?.replace('gpg: key ', '')
+      .split(':')
+      .shift();
   }
-  const keyFileName = upath.join(os.tmpdir() + '/git-private.key');
-  await fs.outputFile(keyFileName, gitPrivateKey!);
-  const { stdout, stderr } = await exec(`gpg --import ${keyFileName}`);
-  logger.debug({ stdout, stderr }, 'Private key import result');
-  keyId = `${stdout}${stderr}`
-    .split(newlineRegex)
-    .find((line) => line.includes('secret key imported'))
-    ?.replace('gpg: key ', '')
-    .split(':')
-    .shift();
-  await fs.remove(keyFileName);
 }
 
-export async function writePrivateKey(): Promise<void> {
-  if (!gitPrivateKey) {
-    return;
+class SSHKey extends PrivateKey {
+  protected readonly gpgFormat = 'ssh';
+
+  protected async importKey(): Promise<string | undefined> {
+    if (await this.hasPassphrase()) {
+      throw new Error('SSH key must have an empty passhprase');
```

Comment:
Why is this line not covered? There is a [test case](https://github.com/renovatebot/renovate/pull/29550/files#diff-7c56595536556edce447a7b810a258d538dfb69b20aa61212cd7edd2148e727aR76-R90) ...

Reply:
Never mind, my bad. I've fixed it and force-pushed (since there hasn't been a review yet).